From fe7e9cb06733d2ac0c40179d5adbc47fbe6f8c4f Mon Sep 17 00:00:00 2001
From: Sam Mortimer <sam@mortimer.me.uk>
Date: Tue, 10 Jan 2017 22:21:45 -0800
Subject: [PATCH 1/1] Reset all package signatures on boot

Can be used to make a special build that resets all
package signatures without wiping any data.

Change-Id: Iff65f2ed3c9d0f80be7221ff90803eda4732cac0
---
 .../java/com/android/server/pm/InstallPackageHelper.java | 9 ++++++++-
 .../com/android/server/pm/PackageManagerService.java     | 7 +++++++
 2 files changed, 15 insertions(+), 1 deletion(-)

diff --git a/services/core/java/com/android/server/pm/InstallPackageHelper.java b/services/core/java/com/android/server/pm/InstallPackageHelper.java
index f449126af0f97..e2800969a96c8 100644
--- a/services/core/java/com/android/server/pm/InstallPackageHelper.java
+++ b/services/core/java/com/android/server/pm/InstallPackageHelper.java
@@ -1498,7 +1498,14 @@ private void preparePackageLI(InstallRequest request) throws PrepareFailure {
                 final KeySetManagerService ksms = mPm.mSettings.getKeySetManagerService();
                 final SharedUserSetting signatureCheckSus = mPm.mSettings.getSharedUserSettingLPr(
                         signatureCheckPs);
-                if (ksms.shouldCheckUpgradeKeySetLocked(signatureCheckPs, signatureCheckSus,
+                if (mPm.mResetSignatures) {
+                    Slog.d(TAG, "resetting signatures on package " + parsedPackage.getPackageName());
+                    signatureCheckPs.setSigningDetails(parsedPackage.getSigningDetails());
+                    if (signatureCheckSus != null) {
+                        signatureCheckSus.signatures.mSigningDetails =
+                                parsedPackage.getSigningDetails();
+                    }
+                } else if (ksms.shouldCheckUpgradeKeySetLocked(signatureCheckPs, signatureCheckSus,
                         scanFlags)) {
                     if (!ksms.checkUpgradeKeySetLocked(signatureCheckPs, parsedPackage)) {
                         throw new PrepareFailure(INSTALL_FAILED_UPDATE_INCOMPATIBLE, "Package "
diff --git a/services/core/java/com/android/server/pm/PackageManagerService.java b/services/core/java/com/android/server/pm/PackageManagerService.java
index 1ca4b2bddd62e..89e66ddcd4106 100644
--- a/services/core/java/com/android/server/pm/PackageManagerService.java
+++ b/services/core/java/com/android/server/pm/PackageManagerService.java
@@ -365,6 +365,8 @@ public class PackageManagerService implements PackageSender, TestUtilityService
 
     static final boolean HIDE_EPHEMERAL_APIS = false;
 
+    private static final boolean RESET_ALL_PACKAGE_SIGNATURES_ON_BOOT = true;
+
     static final String PRECOMPILE_LAYOUTS = "pm.precompile_layouts";
 
     private static final int RADIO_UID = Process.PHONE_UID;
@@ -710,6 +712,7 @@ public class PackageManagerService implements PackageSender, TestUtilityService
     final ProtectedPackages mProtectedPackages;
 
     private boolean mFirstBoot;
+    boolean mResetSignatures;
 
     final boolean mIsEngBuild;
     private final boolean mIsUserDebugBuild;
@@ -2286,6 +2289,8 @@ public Set<String> getInstallConstraintsAllowlist() {
             mShouldStopSystemPackagesByDefault = mContext.getResources()
                     .getBoolean(R.bool.config_stopSystemPackagesByDefault);
 
+            mResetSignatures = RESET_ALL_PACKAGE_SIGNATURES_ON_BOOT;
+
             final int[] userIds = mUserManager.getUserIds();
             PackageParser2 packageParser = mInjector.getScanningCachingPackageParser();
             mOverlayConfig = mInitAppsHelper.initSystemApps(packageParser, packageSettings, userIds,
@@ -2293,6 +2298,8 @@ public Set<String> getInstallConstraintsAllowlist() {
             mInitAppsHelper.initNonSystemApps(packageParser, userIds, startTime);
             packageParser.close();
 
+            mResetSignatures = false;
+
             mRequiredVerifierPackages = getRequiredButNotReallyRequiredVerifiersLPr(computer);
             mRequiredInstallerPackage = getRequiredInstallerLPr(computer);
             mRequiredUninstallerPackage = getRequiredUninstallerLPr(computer);
